`Development`: Validate example solution publication date

akesfeden opened the pull request:
<!-- Thanks for contributing to Artemis! Before you submit your pull request, please make sure to check the following boxes by putting an x in the [ ] (don't: [x ], [ x], do: [x]) -->
<!-- If your pull request is not ready for review yet, create a draft pull request! -->

### Checklist
#### General
<!-- You only need to choose one of the first two check items: Generally, test on the test servers. -->
<!-- If it's only a small change, testing it locally is acceptable and you may remove the first checkmark. If you are unsure, please test on the test servers. -->
- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] This was confirmed by another developer on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.ase.in.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/development-process.html#naming-conventions-for-github-pull-requests).
#### Server
- [x] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I implemented the changes with a good performance and prevented too many database calls.
#### Client
- [x] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client/).
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client-tests/).
- [x] I added multiple screenshots/screencasts of my UI changes.
- [x] I translated all newly inserted strings into English and German.

### Motivation and Context
<!-- Why is this change required? What problem does it solve? -->
<!-- If it fixes an open issue, please link to the issue here. -->
Example Solution Publication Date did not have a min date constraint, this could lead to potential mistakes like publishing a solution before the exercise is over or even before it begins.
### Description
<!-- Describe your changes in detail -->
Example Solution Publication Date of an exercise should be after Release Date and Due Date if they are set. It is allowed to be before the Due Dateif the exercise is optional, i.e. not included in grading. It can also be empty. This validation is made both at client and the server.

Also, at client side the user is now warned if they set the Example Solution Publication Date before the Due Date for an optional exercise in order to prevent early publications of the solution accidentally.

Thanks to @Hialus we noticed that date validation is not performed at all for programming exercises on updates, it is only done during inserts. This is also fixed in this PR.

### Steps for Testing
<!-- Please describe in detail how the reviewer can test your changes. -->
Prerequisites:
- 1 Instructor
- 1 Programming Exercise
- 1 Text Exercise
- 1 Modeling Exercise
- 1 File Upload Exercise

1. Log in to Artemis
2. Navigate to Course Administration
3. For each exercise type above, try the following:
4. Save the exercise with an empty Example Solution Publication Date when release date or due date is set and see it succeed.
5. Save the exercise with an Example Solution Publication Date after the due date and see it succeed.
6. Save the exercise with empty Example Solution Publication Date, release date, due date and see it succeed.
7. Unset the due date and save the exercise with an Example Solution Publication Date after the release date and see it succeed.
8. Set `Should this exercise be included in the course score calculation?` to either `Yes`  or `Bonus`, then try to save the exercise with an Example Solution Publication Date before due date and see it fail with a date validation error.
9. Set `Should this exercise be included in the course score calculation?` to `No`, then try to save the exercise with an Example Solution Publication Date before due date (after release date if set) and see it succeed with a yellow warning alert shown after save button is pressed.
10. Unset the due date and try to save the exercise with an Example Solution Publication Date before release date and see it fail with a date validation error.
### Review Progress
<!-- Each Pull Request should be reviewed by at least two other developers. The code as well as the functionality (= manual test) needs to be reviewed. -->
<!-- The reviewer or author check the following boxes depending on what was reviewed or tested. All boxes should be checked before merge. -->
<!-- You can add additional checkboxes if it makes sense to only review parts of the code or functionality. -->
<!-- When changes are pushed, uncheck the affected boxes. (Not all changes require full re-reviews.) -->

#### Code Review
- [x] Review 1
- [ ] Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

### Test Coverage
<!-- Please add the test coverages for all changed files here. You can see this when executing the tests locally (see build.gradle and package.json) or when looking into the corresponding Bamboo build plan. -->
<!-- Lines are the main reference but a significantly lower branch percentage can indicate missing edge cases in the tests. -->
| Class/File | Branch | Line |
|------------|-------:|-----:|
| BaseExercise.java | 89% | 98% |
| exercise.model.ts | 97.14% | 98.75% |
| file-upload-exercise-update.component.ts | 77.5% | 85% |
| modeling-exercise-update.component.ts | 80.95% | 91.26% |
| programming-exercise-lifecycle.component.ts | 77.77% | 94.59% |
| exercise.service.ts | 82.2% | (did not add tests for just http calls ) 73.15% |
| exercise.utils.ts | 86.45% | 90.9% |
| text-exercise-update.component.ts | 84.41% | 91.39% |
| Exercise.java | 82% | 93% |
| ProgrammingExerciseService.java | 81% | 92% |
| ProgrammingExerciseResource.java | 81% | 90% |

### Screenshots
<!-- Add screenshots to demonstrate the changes in the UI. -->
<!-- Create a GIF file from a screen recording in a docker container https://toub.es/2017/09/11/high-quality-gif-with-ffmpeg-and-docker/ -->

![image](https://user-images.githubusercontent.com/18427209/169002290-3e0794c3-281f-4306-845f-b08466130015.png)

![image](https://user-images.githubusercontent.com/18427209/168967299-0f53440b-ae40-48cd-8e92-70112c7f0a9a.png)

JohannesStoehr reviewed (changes requested):
I managed to save this programming exercise on ts1 (dueDate after solution publication)
![Bildschirmfoto 2022-05-02 um 13 31 13](https://user-images.githubusercontent.com/38322605/166227095-f548a7c3-7d4c-4911-a404-c2b4b31bb27b.png)

Marvin-Vitcu reviewed (changes requested):
i agree with @JohannesStoehr 

- test 8 fails: you can create an "Example Solution Publication Date" before the Due Date
- test 9 fails: you can create an "Example Solution Publication Date" before the Release Date
- The order in which you enter the dates matter: e.g. if you enter the "Example Solution Publication Date" first and then the Release Date or the Due Date there won't be any restrictions. But if you first enter the Release Date or the Due Date you cannot set the "Example Solution Publication Date" before that

Hialus commented:
I might have found the reason for the mentioned issues. It appears that `Exercise::validateGeneralSettings` which calls the `Exercise::validateDates` method is only called when creating a programming exercise, but not when updating it.
For other exercise types this is not the case and the `validateGeneralSettings` is called when creating and updating the exercise.

akesfeden commented:
Thanks for the diagnosis @Hialus, I added validation checks to the endpoints and also fixed the validation on client side as well.

SimonPaul02 reviewed (changes requested):
Tested together with @yuling-sun on Ts1, everything works for programming exercises. But for the rest, you can't save the exercise if you have an empty due date and an empty release date

![image](https://user-images.githubusercontent.com/68793520/166967093-406da3d8-331b-4341-835f-cde5e8652e7f.png)

bottbenj commented:
There are use-cases for having the solution be available before the the due date of an exercise:

For example we want to publish the solution for our tutorial exercises on Friday afternoon but the due date should only be on Monday morning. (We don't simply move the due date, because that is when the homework is due and if the due dates were different the tutorial and homework exercises for a week would appear in different groups on the course page)

Thus we would ask that for at least optional exercises (that don't count towards the grade anyways) it remains possible to set the solution publication date before the due date.

Dominik-Weinzierl commented on src/main/java/de/tum/in/www1/artemis/domain/BaseExercise.java:
```diff
@@ -236,6 +237,20 @@ private static boolean isValidAssessmentDueDate(ZonedDateTime releaseDate, Zoned
+    /**
+     * This method is used to validate the exampleSolutionPublicationDate of an exercise. An exampleSolutionPublicationDate is valid if it is after the releaseDate and dueDate.
+     * Any given exampleSolutionPublicationDate is valid if releaseDate and dueDate are not set.
+     * exampleSolutionPublicationDate is valid if it is not set.
```
You could think about adding `@param` to your comment. 
And the following method could be:
`exampleSolutionPublicationDate == null || (isBeforeAndNotNull(dueDate, exampleSolutionPublicationDate) && isBeforeAndNotNull(releaseDate, exampleSolutionPublicationDate));`. 

But this is only a suggestion. In my opinion, both ways are fine.

akesfeden replied:
Thanks for the suggestions Dominik! I would prefer to keep both as it is right now due to being consistent with the rest of the files. `@param` is not used for the same date type parameters in other validation methods, probably because it would just repeat the name of the date parameter.

And null checks are handled with a return early pattern in the other validation methods as well which enables the reader capture the date constraints with a quick glance in my opinion.

Dominik-Weinzierl replied:
Good and valid points :+1: thank you for your explanation.

Dominik-Weinzierl reviewed (commented):
Code looks good to me - only a minor suggestion for you to think about :+1:

willergeorg reviewed (commented):
While testing I ran in the same problem @SimonPaul02 and @yuling-sun already reported. Please specify your testing steps since at least step 7 is not feasible for any other exercise type except programming exercises.

akesfeden commented:
@SimonPaul02 I made some fixes, it should be possible save the exercise if you have an empty due date and an empty release date now.

akesfeden commented:
@bottbenj 

> There are use-cases for having the solution be available before the the due date of an exercise:
> 
> For example we want to publish the solution for our tutorial exercises on Friday afternoon but the due date should only be on Monday morning. (We don't simply move the due date, because that is when the homework is due and if the due dates were different the tutorial and homework exercises for a week would appear in different groups on the course page)
> 
> Thus we would ask that for at least optional exercises (that don't count towards the grade anyways) it remains possible to set the solution publication date before the due date.

This case is now handled as well, the users are allowed to save in that case but they see a warning alert just in case. I updated the description and the test steps.

akesfeden commented:
@willergeorg I made some changes in the validation constraints and it seems testing step 7 is working now. Please let me know it is still not working for you and I can examine further.

Strohgelaender commented on src/main/java/de/tum/in/www1/artemis/domain/BaseExercise.java:
```diff
@@ -213,8 +214,12 @@ public void validateDates() {
-        // at least one is set, so we have to check the two possible errors
-        boolean areDatesValid = isBeforeAndNotNull(getReleaseDate(), getDueDate()) && isValidAssessmentDueDate(getReleaseDate(), getDueDate(), getAssessmentDueDate());
+
+        IncludedInOverallScore includedInOverallScore = this instanceof Exercise thisAsExercise ? thisAsExercise.getIncludedInOverallScore() : null;
```
I don't like this change. `BaseExercise` is a superclass of exercise, this line turns around the relation.

A better solution might be to make `isValidExampleSolutionPublicationDate` non-static and abstract / overwrite the implementation in the subclass to respect the `includedInOverallScore`-value here.

akesfeden replied:
Yes I wasn't comfortable with that line very much either. That sounds better thank you. What do you think about moving `validateDates` to `Exercise.java` and making `isValidExampleSolutionPublicationDate` etc. `protected` instead?

akesfeden replied:
I pushed the changes I mentioned for review but if you think the way you mentioned would be better I can change it.

alaraozdenler reviewed (changes requested):
Tested on TS3, step 9 didn't work for me as I was not allowed to set an "Example Solution Publication Date" before the "Due Date" for a programming exercise after I set the "Set Should this exercise be included in the course score calculation?" to "No"

akesfeden commented:
Thanks @alaraozdenler, I fixed that issue.

bottbenj commented on src/main/webapp/app/exercises/programming/shared/lifecycle/programming-exercise-lifecycle.component.ts:
```diff
@@ -82,4 +86,20 @@ export class ProgrammingExerciseLifecycleComponent implements OnInit {
+                this.exercise.dueDate != undefined
+                    ? 'artemisApp.programmingExercise.timeline.alertNewExampleSolutionPublicationDateAsDueDate'
+                    : 'artemisApp.programmingExercise.timeline.alertNewExampleSolutionPublicationDateAsReleaseDate';
+            alert(this.translator.instant(message));
```
this should probably be an alertservice info popup instead of a js alert to not just stop everything

akesfeden replied:
`updateDueDate` method above that one also used `alert()` so I wanted to make it consistent with it. Also, since the date is changed after the user closes the alert box, this gives them a chance to look at the previous value of the date that will be changed. Being able to do a last manual check can be convenient for the user  in my opinion.

https://github.com/ls1intum/Artemis/blob/cda2dbdf339f98e3f5cdcfb66cd5a66c9b57be43/src/main/webapp/app/exercises/programming/shared/lifecycle/programming-exercise-lifecycle.component.ts#L78-L88

bottbenj reviewed (commented):
Other than that the warning that the example solution publication is before the due date comes only after saving the exercise which may be too late to change if that was not intended, for programming exercises it could be shown in the footer, tho this is not terribly important.

Using the `Edit Selected` on the management exercise list i was able to save 
![image](https://user-images.githubusercontent.com/44437514/169414448-eba18bbd-96fe-4c5c-938a-37114f7a50e7.png) which when loading the edit page gives two alerts that the solution date was set to the due date but doesn't actually do that.

On the `Edit Selected` dialog the datetime picker for the solution date also doesn't show up if the due date is set, instead somewhat softlocking the page until i reload manually.

akesfeden commented:
@bottbenj 

> Other than that the warning that the example solution publication is before the due date comes only after saving the exercise which may be too late to change if that was not intended, for programming exercises it could be shown in the footer, tho this is not terribly important.

This may need more data transfer between components so if it is not very important I'd suggest we evaluate whether adding complexity for that (although not a big one) is worth the benefit in a possible future PR.

> Using the `Edit Selected` on the management exercise list i was able to save ![image](https://user-images.githubusercontent.com/44437514/169414448-eba18bbd-96fe-4c5c-938a-37114f7a50e7.png) which when loading the edit page gives two alerts that the solution date was set to the due date but doesn't actually do that.

I added fix for this and the softlocking issue below at the latest commit.

> On the `Edit Selected` dialog the datetime picker for the solution date also doesn't show up if the due date is set, instead somewhat softlocking the page until i reload manually and that dialog is generally somewhat broken. but that happens on prod as well.

(See above)

> Finally solution publication date and no due date for graded exercises should at least get a warning like before due date for optional exercises.

I added the same warning when the due date is empty for graded an non graded exercises. I think using the same message ("Example solution publication date is before due date") for this case also makes sense because a non-empty example solution publication date eventually arrives unlike an empty due date so it is before the due date.

canberkanar commented:
> "Example solution publication date is before due date")

"_Example solution publication date is before due date_", I think that when we do not set a due date and show this warning message it might be a little confusing to the end-user, but functionality wise there are no problems.

ole-ve reviewed (commented):
Not related to the scope of this PR, but the "Save"- button in modeling exercises is not disabled for date inputs like random strings. The validation fails on the individual form fields, but the button is still enabled.
You can take a look at this file to invalidate a submit button when there are invalid input fields:
https://github.com/ls1intum/Artemis/blob/develop/src/main/webapp/app/exercises/shared/exercise-hint/manage/exercise-hint-update.component.html
`type="submit"` is the relevant part for the save button.

canberkanar reviewed (approved):
Tested on ts1 in the testing session. A quick fix was immediately made to disable the Save button when the publication date is before the release date. Well done! 👍🏻 
![image](https://user-images.githubusercontent.com/67344158/169798577-05297dfd-8a03-4588-978a-96c6d6db9880.png)

Lugitan reviewed (approved):
Tested on TS1. Works as described.

manuelmanso reviewed (approved):
Tested during test session on TS1, works as expected after quickfix during session

minrows reviewed (approved):
Tested in the testing session on TS1, works as described.
